include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -0,0 +1,301 @@
+#else
+		return glm::findMSB(val);
+#endif
+	}
+};
+
+template<>
+struct find_msb_helper<int32_t>
+{
+	static int32_t findMSB(NBL_CONST_REF_ARG(int32_t) val)
+	{
+#ifdef __HLSL_VERSION
+		return spirv::findSMsb(val);
+#else
+		return glm::findMSB(val);
+#endif
+	}
+};
+
+template<int N>
+struct find_msb_helper<vector<uint32_t, N> >
+{
+	static vector<int32_t, N> findMSB(NBL_CONST_REF_ARG(vector<uint32_t, N>) val)
+	{
+#ifdef __HLSL_VERSION
+		return spirv::findUMsb(val);
+#else
+		return glm::findMSB(val);
+#endif
+	}
+};
+
+template<int N>
+struct find_msb_helper<vector<int32_t, N> >
+{
+	static vector<int32_t, N> findMSB(NBL_CONST_REF_ARG(vector<int32_t, N>) val)
+	{
+#ifdef __HLSL_VERSION
+		return spirv::findSMsb(val);
+#else
+		return glm::findMSB(val);
+#endif
+	}
+};
+
+#ifdef __HLSL_VERSION
+template<typename Integer>
+struct find_lsb_helper;
+#else
+// legacy code wouldn't work without it
+template<typename Integer>
+struct find_lsb_helper
+{
+	static int32_t findLSB(NBL_CONST_REF_ARG(Integer) val)
+	{
+#ifdef __HLSL_VERSION
+		return spirv::findILsb(val);
+#else
+		if (is_signed_v<Integer>)
+		{
+			// GLM accepts only integer types, so idea is to cast input to integer type
+			using as_int = typename integer_of_size<sizeof(Integer)>::type;
+			const as_int valAsInt = reinterpret_cast<const as_int&>(val);
+			return glm::findLSB(valAsInt);
+		}
+		else
+		{
+			// GLM accepts only integer types, so idea is to cast input to integer type
+			using as_uint = typename unsigned_integer_of_size<sizeof(Integer)>::type;
+			const as_uint valAsUnsignedInt = reinterpret_cast<const as_uint&>(val);
+			return glm::findLSB(valAsUnsignedInt);
+		}
+#endif
+	}
+};
+#endif
+
+template<>
+struct find_lsb_helper<int32_t>
+{
+	static int32_t findLSB(NBL_CONST_REF_ARG(int32_t) val)
+	{
+#ifdef __HLSL_VERSION
+		return spirv::findILsb(val);
+#else
+		return glm::findLSB(val);
+#endif
+	}
+};
+
+template<>
+struct find_lsb_helper<uint32_t>
+{
+	static int32_t findLSB(NBL_CONST_REF_ARG(uint32_t) val)
+	{
+#ifdef __HLSL_VERSION
+		return spirv::findILsb(val);
+#else
+		return glm::findLSB(val);
+#endif
+	}
+};
+
+template<int N>
+struct find_lsb_helper<vector<int32_t, N> >
+{
+	static vector<int32_t, N> findLSB(NBL_CONST_REF_ARG(vector<int32_t, N>) val)
+	{
+#ifdef __HLSL_VERSION
+		return spirv::findILsb(val);
+#else
+		return glm::findLSB(val);
+#endif
+	}
+};
+
+template<int N>
+struct find_lsb_helper<vector<uint32_t, N> >
+{
+	static vector<int32_t, N> findLSB(NBL_CONST_REF_ARG(vector<uint32_t, N>) val)
+	{
+#ifdef __HLSL_VERSION
+		return spirv::findILsb(val);
+#else
+		return glm::findLSB(val);
+#endif
+	}
+};
+
+template<typename Integer>
+struct find_msb_return_type
+{
+	using type = int32_t;
+};
+template<typename Integer, int N>
+struct find_msb_return_type<vector<Integer, N> >
+{
+	using type = vector<int32_t, N>;
+};
+template<typename Integer>
+using find_lsb_return_type = find_msb_return_type<Integer>;
+
+template<typename T, typename U>
```

Comment:
add `NBL_STRUCT_CONSTRAINABLE` instead of closing brace, so that you can have constrained partial specs later